### PR TITLE
add Pallas and Jubjub ciphersuites and FROST support

### PR DESCRIPTION
**Pull request by conradoplg**
Part of https://github.com/ZcashFoundation/frost/issues/121 along with https://github.com/ZcashFoundation/frost/pull/152

### CODE REVIEW

File: src/frost_redjubjub.rs [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+    fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self::Scalar {
+        Self::Scalar::random(rng)
+    }
+
+    fn serialize(scalar: &Self::Scalar) -> Self::Serialization {
+        scalar.to_repr()
+    }
+
+    fn little_endian_serialize(scalar: &Self::Scalar) -> Self::Serialization {
```
[the remark this answers is not in the document]
Comment (conradoplg) [reply src/frost_redjubjub.rs:55]: Good catch, I change it to `to_bytes()` in 90b5d4369bd044b369e28b5bf2ef27d2f682a781

File: src/frost_redpallas.rs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+    fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self::Scalar {
+        Self::Scalar::random(rng)
+    }
+
+    fn serialize(scalar: &Self::Scalar) -> Self::Serialization {
+        scalar.to_repr()
+    }
+
+    fn little_endian_serialize(scalar: &Self::Scalar) -> Self::Serialization {
```
[the remark this answers is not in the document]
Comment (conradoplg) [reply src/frost_redpallas.rs:59]: Good catch, in this case there is no other option so I added a comment in 90b5d4369bd044b369e28b5bf2ef27d2f682a781

**Review by conradoplg — COMMENTED** [review]
Thanks @upbqdn !

**Comment by conradoplg** [comment]
@upbqdn I had to update some code to match the latest changes in `frost-rerandomized`, and to also fix CI

**Review by upbqdn — DISMISSED** [review]
I think this PR looks good and can be merged, but I'm going to request changes to postpone merging so that @dconnolly can still have a look (I didn't want the PR to disappear suddenly). I didn't find a more suitable way to do this.

**Comment by conradoplg** [comment]
One thing I'm not sure is the structure, I'm not happy with the `frost_redpallas.rs` and `frost_redjubjub.rs`.

Some other options:

- Use `./frost/red(pallas|jubjub).rs`
- Make this a workspace and add `frost-red(pallas|jubjub)` crates, to match other the `frost` repo (e.g. `frost-ristretto255`)

@dconnolly any thoughts?

**Comment by dconnolly** [comment]
> One thing I'm not sure is the structure, I'm not happy with the `frost_redpallas.rs` and `frost_redjubjub.rs`.
> 
> Some other options:
> 
> * Use `./frost/red(pallas|jubjub).rs`
> * Make this a workspace and add `frost-red(pallas|jubjub)` crates, to match other the `frost` repo (e.g. `frost-ristretto255`)
> 
> @dconnolly any thoughts?

<many months later>

This is an interesting question. In my mind, doing reddsa signatures via FROST or via a classic single party are somewhat parallel, especially in the keygen parts; but also, they are two ways to create 'the same' signatures, so they are one thing. I think I lean towards 

> * Use `./frost/red(pallas|jubjub).rs`

especially as `reddsa` is becoming a collection, rather than cleanly siloed libraries

**Review by dconnolly — COMMENTED** [review]
Good, I am making suggestions to the `frost` generic stuff that we should pull in here

**Comment by conradoplg** [comment]
> This is an interesting question. In my mind, doing reddsa signatures via FROST or via a classic single party are somewhat parallel, especially in the keygen parts; but also, they are two ways to create 'the same' signatures, so they are one thing. I think I lean towards
> 
> > * Use `./frost/red(pallas|jubjub).rs`
> 
> especially as `reddsa` is becoming a collection, rather than cleanly siloed libraries

Thanks!

Moved to `frost` folder in 6a72035
Also fixed a compiling issue in 6baedcf7d129e334b925640a35c571ecb6ed761f

File: src/frost/redpallas.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+use frost_rerandomized::{
+    frost_core::{frost, Ciphersuite, Field, FieldError, Group, GroupError},
+    RandomizedParams,
+};
+
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{hash::HStar, orchard, private::Sealed};
+
+/// An error.
```
Comment (upbqdn) [inline src/frost/redpallas.rs:19]: ###### Optional

```suggestion
/// An error type for the FROST(Pallas, BLAKE2b-512) ciphersuite.
```

File: src/frost/redjubjub.rs [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+use frost_rerandomized::{
+    frost_core::{frost, Ciphersuite, Field, FieldError, Group, GroupError},
+    RandomizedParams,
+};
+
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{hash::HStar, private::Sealed, sapling};
+
+/// An error.
```
Comment (upbqdn) [inline src/frost/redjubjub.rs:18]: ###### Optional

```suggestion
/// An error type for the FROST(Jubjub, BLAKE2b-512) ciphersuite.
```

**Review by upbqdn — APPROVED** [review]
I went over the code again, looks great. Two optional suggestions.

**Pull request merged**